### PR TITLE
Added oauth introspection endpoint

### PR DESCRIPTION
**Pull request by IndunilRathnayake**

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/pom.xml [hunk 1]
```diff
@@ -97,6 +97,17 @@
             <artifactId>org.wso2.carbon.identity.oidc.session</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>org.codehaus.jettison</groupId>
+            <artifactId>jettison</artifactId>
+            <version>1.3.7</version>
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.endpoint/pom.xml]: can't we move the version to parent pom ?

**Comment by IndunilRathnayake** [comment]
1)  
Get a valid token: 
curl -v -X POST --basic -u W80BsTkU7u02ITYQWlFNgSq9MuAa:yPyzQsrj8t1mCiIx_6tYOzu6pCIa -H "Content-Type: application/x-www-form-urlencoded;charset=UTF-8" -k -d "grant_type=client_credentials" https://localhost:9443/oauth2/token 

Response:
{"token_type":"Bearer","expires_in":3600,"access_token":"fbc4e794-23db-3394-b1e5-f2c3e511d01f"}

Validate the token:
curl -k -H 'Content-Type: application/x-www-form-urlencoded' -X POST --data 'token=fbc4e794-23db-3394-b1e5-f2c3e511d01f' https://localhost:9443/oauth2/introspect 

{"exp":1464161608,"username":"admin@carbon.super","active":true,"token_type":"Bearer","client_id":"W80BsTkU7u02ITYQWlFNgSq9MuAa","iat":1464158008}



2)
Get a valid token with a scope: 
curl -v -X POST --basic -u W80BsTkU7u02ITYQWlFNgSq9MuAa:yPyzQsrj8t1mCiIx_6tYOzu6pCIa -H "Content-Type: application/x-www-form-urlencoded;charset=UTF-8" -k -d "grant_type=client_credentials&scope=test1 test2" https://localhost:9443/oauth2/token 

{"scope":"test1 test2","token_type":"Bearer","expires_in":3153,"access_token":"b33e297e-4dec-31c6-a5b0-dd9d9b5e3e3d"}

Validate the token:
curl -k -H 'Content-Type: application/x-www-form-urlencoded' -X POST --data 'token=b33e297e-4dec-31c6-a5b0-dd9d9b5e3e3d' https://localhost:9443/oauth2/introspect  

{"exp":1464161560,"username":"admin@carbon.super","scope":"test1 test2","active":true,"token_type":"Bearer","client_id":"W80BsTkU7u02ITYQWlFNgSq9MuAa","iat":1464157960}

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/OAuth2IntrospectionEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+        if(tokenTypeHint == null) {
+            tokenTypeHint = DEFAULT_TOKEN_TYPE_HINT;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Token type hint: " + tokenTypeHint);
+        }
+
+        if (token == null || token.trim().length() == 0) {
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/OAuth2IntrospectionEndpoint.java]: use StringUtils

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponse.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.oauth.endpoint.introspection;
+
+public final class IntrospectionResponse {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponse.java]: Missing class level comment.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponseBuilder.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+import java.util.Map;
+
+/**
+ * 
+ * this class is responsible for building the introspection response.
+ * 
+ */
+public class IntrospectionResponseBuilder {
+
+    private Map<String, Object> parameters = new HashMap<String, Object>();
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponseBuilder.java]: Can use java 7 diamond operator.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponseBuilder.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    private boolean isActive = false;
+
+    /**
+     * build the introspection response.
+     * 
+     * @return
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponseBuilder.java]: Incomplete method comments.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponseBuilder.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    /**
+     * 
+     * @param isActive
+     * @return
+     */
+    public IntrospectionResponseBuilder setActive(boolean isActive) {
+	parameters.put(IntrospectionResponse.ACTIVE, isActive);
+	if (!isActive) {
+	    // if the token is not active we do not want to return back the expiration time.
+	    if (parameters.containsKey(IntrospectionResponse.EXP)) {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/IntrospectionResponseBuilder.java]: Unnecessary if check. HashMap will not throw an NPE if the entry is not there.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/webapp/WEB-INF/web.xml [hunk 7]
```diff
@@ -80,6 +80,7 @@
         <cookie-config>
             <secure>true</secure>
         </cookie-config>
+        <session-timeout>60</session-timeout>
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/webapp/WEB-INF/web.xml]: What is the need of above change?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/OAuth2IntrospectionEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+@Path("/introspect")
+@Consumes({ MediaType.APPLICATION_FORM_URLENCODED })
+@Produces(MediaType.APPLICATION_JSON)
+public class OAuth2IntrospectionEndpoint {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/OAuth2IntrospectionEndpoint.java]: Missing class level comment.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/OAuth2IntrospectionEndpoint.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+        if (introspectionResponse.getError() != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("The error why token is made inactive: " + introspectionResponse.getError());
+            }
+            return Response.status(Response.Status.OK).entity("{'active':false}").build();
+        }
+
+        IntrospectionResponseBuilder respBuilder = new IntrospectionResponseBuilder().setActive(introspectionResponse.isActive())
+                .setNotBefore(introspectionResponse.getNbf()).setScope(introspectionResponse.getScope()).setUsername(introspectionResponse.getUsername())
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/introspection/OAuth2IntrospectionEndpoint.java]: Check for 120 columnwidth.

File: pom.xml [hunk 10]
```diff
@@ -249,6 +254,7 @@
         <opensaml.version>2.6.4</opensaml.version>
         <opensaml2.wso2.version>2.6.4.wso2v2</opensaml2.wso2.version>
         <opensaml2.wso2.osgi.version.range>[2.6.0,3.0.0)</opensaml2.wso2.osgi.version.range>
+        <org.codehaus.jettison.version>1.3.7</org.codehaus.jettison.version>
```
Comment (omindu) [inline pom.xml]: Is it possible to use GSON as it is the commonly used JSON library.

**Comment by omindu** [comment]
Merging this for the time being since this is urgent. Please make changes as suggested and send another PR.

**Comment by IndunilRathnayake** [comment]
Resolved all comments in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/80

**Pull request merged**